### DISCUSSION TITLE
Questions on VAD (sample_rate, length of channels, device)

### DISCUSSION THREAD

**Discussion by kaen2891** [speechbrain/speechbrain#1649] [Q&A]
I have questions about VAD pre-trained model.

1. Sample rate
Assume that given the sample rate of random wav file as not 16kHz, how can I change to use for VAD pre-trained model (`get_speech_prob_file` function)? It seems like `get_speech_prob_file` function only allows to get the file path of the wav file. 

![image](https://user-images.githubusercontent.com/46586785/143815918-08e4134f-31de-4f42-a733-cd47d0fd7c35.png)

Because the `get_speech_prob_file` function allows only the file path.

Do I have to load first and resample that wav file, and save, and load it again?

2. Length of channels
If the wav data has [2, T], how could I use the `get_speech_prob_file` function? It seems like only allowing the [1, T] of the wav file. Do I have to load first and cut as [1, T] and save, ... and load it again?
![image](https://user-images.githubusercontent.com/46586785/143868992-5e85d5ce-c112-4fe6-ab63-1fd7a3a839fc.png)

Assume that I can't use ```torchaudio```'s ```save``` function by changing the ```2 channel``` wav to 1 channel, how can I solve it?

3. Device
I wanna use the VAD pre-trained model as Cuda device, but I got errors when I load the wav file as below:
![image](https://user-images.githubusercontent.com/46586785/143815484-f168d843-a34b-49fb-a470-28271c521cae.png)

Because the input format of `get_speech_prob_file` is only available at `filepath (e.g., /home/data.wav)`. So I don't know how to make this file on Cuda available.

How can use this as cuda? Only can perform on CPU?

**Comment by kaen2891** [comment]
The most hurry things that I want to handle are the ```number 2``` that is the length of channels. If I cannot use the torchaudio.save in a limited environment, how can I use the ```2 channels of wav file (stereo)```? Could you help me? @mravanelli

**Comment by mravanelli** [comment]
Hi @kaen2891,
thank you for the interest in SpeechBrain.
Let me reply to your questions:

1.  **Sample rate**
The supported sample rate depends on the one used for training the model. In the [vad-crdnn-libriparty case](https://huggingface.co/speechbrain/vad-crdnn-libriparty), the sample rate is 16 kHz.  If your signal is at a higher sampling rate (e.g., 44.1 kHz or 48 kHz), you can downsample it to 16 kHz before feeding it into the VAD. If your sampling rate is smaller (e.g., 8 kHz) you can upsample at 16 kHz.
The upsampling is a bit more critical because you lose some information but normally works well anyway. To resample a signal,  you can use torchaudio or sox.  So yes, you have to load, resample, and save the resampled version on disk before using the VAD. As an alternative, you can do the resampling on the fly right after reading each chunk of signal, [after this line](https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/pretrained/interfaces.py#L1071).

2. **Length of channels**
The system expects single-channel audio. You can save the signal as single-channel audio or you can select the channel right after reading the signal [here](https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/pretrained/interfaces.py#L1071).

3. **Device**
The device mismatch issue is already solved in the latest dev branch. Make sure you have it installed:

```
git clone https://github.com/speechbrain/speechbrain/
cd speechbrain
pip install -e .
```

The fix will appear in the main branch (i.e. the one that you use with pip install) when we will release the next version of speechbrain (which will be issued soon).

**Comment by kaen2891** [comment]
Thanks for replying, @mravanelli 

I solved the problem of 1 and 2 thanks to your guides.

But I also got an error with ```VAD``` that is installed with the git clone version.

![image](https://user-images.githubusercontent.com/46586785/144015023-41c59fd3-05c9-4183-8028-2fc4fd6db881.png)


I don't know how to solve this.

The input wav file is 16kHz, mono channel.

**Comment by mravanelli** [comment]
Hi,
my feeling is that your input signal is not in the expected format (you might need to transpose the input signal, unsqueeze it,or, something like that). 

You can try to double check it in this way:

1. Try to run the example provided in the [HF repository](https://huggingface.co/speechbrain/vad-crdnn-libriparty) with the provided signal example (see code at the end of this message). Do it work?

2. If it works, I suggest to make sure the dimensionality of your signal is compatible with the format used in the provided example (see` pretrained_model_checkpoints/example_vad.wav`). You can check the dimensionality of the provided example with torchaudio in this way:

```
import torchaudio
sig, fs = torchaudio.load('example_vad.wav')
sig.shape
torch.Size([1, 4794408])

```
3. You can run the code above with your signal to make sure the format is [1, time_steps]. If not, change the format of your signal ti match the expected one (using operations like transpose, unsqueeze, etc)



**Code Snippet with the provided example:**
```
from speechbrain.pretrained import VAD
VAD = VAD.from_hparams(source="speechbrain/vad-crdnn-libriparty", savedir="pretrained_models/vad-crdnn-libriparty")

# 1- Let's compute frame-level posteriors first
audio_file = 'pretrained_model_checkpoints/example_vad.wav'
prob_chunks = VAD.get_speech_prob_file(audio_file)

# 2- Let's apply a threshold on top of the posteriors
prob_th = VAD.apply_threshold(prob_chunks).float()

# 3- Let's now derive the candidate speech segments
boundaries = VAD.get_boundaries(prob_th)

# 4- Apply energy VAD within each candidate speech segment (optional)

boundaries = VAD.energy_VAD(audio_file,boundaries)

# 5- Merge segments that are too close
boundaries = VAD.merge_close_segments(boundaries, close_th=0.250)

# 6- Remove segments that are too short
boundaries = VAD.remove_short_segments(boundaries, len_th=0.250)

# 7- Double-check speech segments (optional).
boundaries = VAD.double_check_speech_segments(boundaries, audio_file,  speech_th=0.5)
```

**Comment by mravanelli** [comment]
@kaen2891, is everything fine now?
